### ISSUE TITLE
Can I make {.async.} and {.raises:[].} works together?

### ISSUE THREAD

**Issue by loongw** [nim-lang/Nim#6639]
```nim
import asyncdispatch

proc p() {.async, raises:[] .} = discard # Error: can raise an unlisted exception
```

**Comment by Austaras** [comment]
any update or workaround?

**Comment by dom96** [comment]
Sadly no. If someone wants to dive into the `async` macro though, feel free, I don't think fixing this should be too much trouble.

**Comment by alehander92** [comment]
@dom96 how do you think this can be fixed? i tried to think of stuff in https://irclogs.nim-lang.org/19-03-2020.html#08:47:00 but i couldnt realize what can be done

**Comment by alehander92** [comment]
also https://irclogs.nim-lang.org/19-03-2020.html#08:16:16 : combine with an AsyncError (maybe FutureError is what i need)

**Comment by alehander92** [comment]
(what i mean mostly is : the resulting effects are usually the effects of the actual call + the async loop effects like OS errors etc)

**Comment by dom96** [comment]
> the async loop effects like OS errors etc

Not really, if an OS error happens when you call `recv` then it belongs to the `recv` call, not the async loop. Hence why it should be raised in the `recv`.

AFAIK you just need to make sure that all exceptions in the code that the async macro generates are handled (we should be doing this anyway), then `raises: []` will just work (TM).

**Comment by alehander92** [comment]
well, i guess the loop can always get into some system limit errors (memory/handles/etc) : one possible thing is to just assume that the event loop doesnt raise from external viewpoint and ignore them: it either works or fails (and errors in it are something like a fatal error)

**Comment by arnetheduck** [comment]
chronos supports this: https://github.com/status-im/nim-chronos/pull/251

**Comment by Zectbumo** [comment]
can we get this working so we have a workaround?
```nim
proc p() {.async.} =
  {.cast(raises:[]).}:
    discard # make sure this block doesn't raise anything
```

**Comment by arnetheduck** [comment]
> can we get this working so we have a workaround?

Your workaround does the opposite .. it would be nice to have a feature to force raises checking locally, not only for async, but it doesn't exist.
